pudb/run.py

```diff
@@ -5,33 +5,44 @@ def main():
     import sys
 
     from optparse import OptionParser
+    # TODO: update cmd string
     parser = OptionParser(
-            usage="usage: %prog [options] SCRIPT-TO-RUN [SCRIPT-ARGUMENTS]")
+            usage="usage: %prog [options] [SCRIPT-TO-RUN|-m MODULE] [ARGUMENTS]")
 
     parser.add_option("-s", "--steal-output", action="store_true"),
+    parser.add_option("-m", "--module", metavar="MODULE",
+                      help="Debug module or package instead of script"),
```

Comment:
Add something here about how this terminates the options list (c.f. `python -h`), since any further arguments should be passed as the argv to the script.

Reply:
Hm, but it doesn't really, does it?
It seems `OptionParser` will keep parsing until it doesn't recognize any other args, I can run a module with `pudb3 -m my_module --pre-run="echo oi" one two three` and OptionParser will give `--pre-run="echo oi"` to pudb and `one two three` to the module

Reply:
I guess you're right. That's how it should behave though. I don't know if optparse or argparse actually support this.

Reply:
Looks like with argparse you can use `REMAINDER`. I don't know if it is possible to do this with optparse. We may need to move over to argparse for this PR. If `-m` does pass all the flags to the module that will be very surprising to people. Does pdb do the right thing here?

Reply:
You can use something like `python -m pudb -m http.server -h` as a test case. It should open the debugger in pudb, and after continuing and exiting, the http.server help should be printed to the terminal. If it shows the pudb help that behavior is wrong.

Reply:
It's quite late for me, so i'll be off now.
If you leave comments, I can look tomorrow.

Reply:
Looks like pdb parses the arguments manually with getopt https://github.com/python/cpython/blob/e822e37946f27c09953bb5733acf3b07c2db690f/Lib/pdb.py#L1661. I don't see why argparse.REMAINDER wouldn't work, but if there is some issue with it, I guess we can use that. That will make it hard if we want to add any other command line arguments in the future, so we should try to use argparse if we can.

Reply:
It doesn't seem to be possible to get the behavior you want for `python -m pudb -m http.server -h` using argparse.
Here is a test script that demonstrates the behavior, which you can try yourself: https://gist.github.com/eliasdorneles/525991dbc996b59ef862a1650fce00eb

If you run it like: `python script.py -m module -h` you will always get the script's help.
In order to get the `-h` passed on to the `script_args` variable, you have to either add another argument before the `-h` or add `--` (which is pretty reasonable to me, that's how many Unix utilities handle these cases of partial parsing).

If this sounds good enough, I can go ahead and change the code to use argparse.

Reply:
Personally, I believe this PR is already useful and making things better, despite of the argument handling bikeshedding. :angel:

Reply:
Presumably that's only because -h is also a flag to pudb. So any other flag would still be passed through. If that's the case, that's better than the current behavior, which steals all `--flags`. 

Requiring for `--` for the corner cases of flags used by pudb is probably fine, though it should be documented in the help. It does mean that we should minimize the single letter flags to pudb to avoid conflicts, but I don't think that's a huge deal. Making it not required would be better but that may be too complicated to implement. 

> Personally, I believe this PR is already useful and making things better, despite of the argument handling bikeshedding. 👼

I don't see it as bikeshedding. -m modules are very often run with command line flags, so users would run into the bug of the current behavior quite quickly I think.

Reply:
Fair enough.

Well, I thought since pudb doesn't have many arguments, another option is to write our own parsing code that does exactly what we want and doesn't introduce any of the limitations of `getopt` and no ambiguities of using `optparse`/`argparse`.

Here is what it would like: https://gist.github.com/eliasdorneles/51fc37dd4910fe07694a30ed6a70aa57

Would it be acceptable to do that here?

Reply:
How is that different from using getopt? I haven't really used getopt before so I don't know much about it.

Reply:
Maybe an easier way is to manually extract everything before `-m` in the sys.argv and pass only that to `parse_args()`. We could even keep using `optparse` with this method (if you don't want to translate to argparse here).

Reply:
Ah, my bad, I had assumed getopt was limited to single-character flags (I had it mixed up with the `getopts` bash builtin).
I will look into the getopt way, as it seems simple enough, less workaround-ish (I'm afraid splitting on `-m` would introduce annoying edge cases, e.g, if it was meant to be passed to a script) and it seems a good idea to mirror the `pdb` way.

Reply:
So I looked into it more closely, and the thing with `getopt` is that it consumes all arguments until it finds a non-argument (much like `parse_known_args` from argparse), and the reason why `python -m pdb -m http.server -h` works with it is that the `-m` is implemented as a boolean flag -- meaning that getopt stops parsing arguments when it see `http.server` because it's the first non-argument.

You can see this by running `python -m pdb -m -c continue http.server -h` <- notice how the -m is given (to activate mode "run as module") and then the module name is given only after given another option recognized by pdb (`-c continue`).

So, if we're okay using this same behavior of `-m` as a boolean flag on pudb, we can then easily implement it with `argparse` and `argparse.REMAINDER`.

I'm going to push a commit implementing that, let me know what do you think.

Reply:
> `python -m pdb -m -c continue http.server -h`

That seems like a corner case that isn't that important. If that's the only thing that is off by that implementation, I am favor of it.

Reply:
Great! I agree!
I've pushed the implementation with argparse, it seems to be working well.
Does it look good now?
Thanks!